### PR TITLE
Fixes #213: Nginx/PHP-FSM environments subject to infinite recursions (internal redirects) - Option B

### LINKED ISSUE #213: Bug: Nginx/PHP-FSM environments subject to infinite recursions (internal redirects)
We've come across a strange occurrence on Nginx/PHP-FSM environments where a combination of the most up to date `silverstripe-trailing-slash` and `sharedraftcontent` versions lead to infinite internal redirects, causing "exceeds memory" fatal errors (or "exceeds time limit" fatal errors, when configured for max memory). The issue did not impact Apache configurations running the identical code base.

Before diving in, I reached out to @wilr (who contributed to #144) for context and suggested direction.

I started just dumping all the data as it was executed from TrailingSlashRedirector.php and comparing the outputs between Nginx/php-fsm (not working) and Apache (working) and eventually found the discrepancy.... 

image
> _Left is Nginx (incorrect), right is Apache (correct). Note that on the left (nginx), the second attempt makes a lookup request for the preview link, when it should have already converted that link to the full slug at the point (as apache correctly did, on the right)._

Ultimately, we isolated it to `ShareDraftController` where it sets up a clean environment internally, sets the URL via `$variables['_SERVER']['REQUEST_URI'] = $url;`, and then saves it to the environment via `Environment::setVariables($variables);`

I printed `Environment::getEnv('REQUEST_URI');` before and again after `Environment::setVariables($variables);` is executed, and sure enough...
image
> _Left is Nginx (incorrect), right is Apache (correct)._

## Solution
Two alternate solutions - both accomplish the same thing in different ways. I'm not as deeply familiar with the `Environment` class and the generally acceptable usages so I'm submitting both for consideration (pick one), though, @wilr and I agree that Option B seems a bit clearer and there might be unknown repercussions to unilaterally overwriting `static::$env` as happens Option A.

1. #211
2. #212 

The end result is the same for both implementations:

image

It really just comes down to whether we use only `Environment::setVariables` (Option A) or a combination of `Environment::setVariables` and `Environment::setEnv` (Option B).

I have to wonder if perhaps `Environment::setVariables` is not being used correctly in the first place?  Because I noticed the source code for `Environment::setVariables` sets `static::$env` variables in a _VERY_ different way compared to `Environment::setEnv`: 

image
 image

## What doesn't work....

I did attempt to replace `Environment::setVariables` entirely with only `Environment::setEnv` calls -- it broke both environments. So, somehow `$GLOBALS` are a necessary component.

## Context / related:

1. https://github.com/axllent/silverstripe-trailing-slash/pull/14
2. #143 
3. #144 - ([related code](https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/144/files#diff-374fb4b6f6a4c240e83c2298460a417f813cabc6544a581e3b45671585773452R163))
4. #214

---

@muskie9 @jsirish @oddnoc @phil-quinn @wilr 

## PRs
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/214
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/211
  OR
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/212

### PR DESCRIPTION
**Pull request by nathanbrauer**
# Fixes 
- #213 

## Info

- I am submitting two pull requests: this one and #211.
- Both solve the infinite recursion issue for Nginx/PHP-FPM builds but in different ways.
- **ONLY ACCEPT ONE!**
- Edits for maintainers are enabled 
- Set to merge into `2` and can be rebased onto `3` without issue

---
@wilr @muskie9

### CODE REVIEW

**Comment by michalkleiner** [comment]
Hi @nathanbrauer, can you please open an issue this PR is trying to solve with your setup info and some replication steps?

**Comment by GuySartorelli** [comment]
I think we'll go with option A, since that seems to be more in line with what's _already_ happening to get this module working in the first place.

**Comment by GuySartorelli** [comment]
Actually - I've changed my mind.
Option A feels like overkill now that I've looked into it further - you're setting _ALL_ of `$_SERVER` into `$env` when you only need these three items.

**Review by GuySartorelli — APPROVED** [review]
Looks good. Confirmed that this fixes the problem with nginx.
I've retargetted this to `2.9` so it can be released as a patch. I reset the commits but made no change to the actual changes the PR is making.

Will merge when CI goes green.

**Comment by GuySartorelli** [comment]
Thanks for contributing! This will be automatically tagged by GitHub actions.

File: src/Controllers/ShareDraftController.php [hunk 1]
```diff
@@ -179,6 +179,11 @@ private function getRenderedPageByURL(string $url): HTTPResponse
 
         Environment::setVariables($variables);
 
+        // This extra step required for PHP-FPM/Nginx implementations
+        Environment::setEnv('REQUEST_URI', $url);
+        Environment::setEnv('REQUEST_METHOD', 'GET');
+        Environment::setEnv('HTTP_USER_AGENT', $variables['_SERVER']['HTTP_USER_AGENT']);
```
Comment (michalkleiner) [inline src/Controllers/ShareDraftController.php:185]: Probably should have `?? ''` if someone does something funny with the nested array (even if unlikely).

Comment (ChloeSartorelli) [reply src/Controllers/ShareDraftController.php:185]: You mean `$variables['_SERVER']['HTTP_USER_AGENT'] ?? ''`?
This is explicitly assigned in line 174 above - I don't think there's an opportunity between that assignment and here for someone to fiddle with the value.

Comment (michalkleiner) [reply src/Controllers/ShareDraftController.php:185]: Of course, I didn't check above the change. All good then 👍

**Pull request merged**